docker/daemon/daemon_src.go

```diff
@@ -334,6 +335,18 @@ func (s *daemonImageSource) GetTargetManifest(digest digest.Digest) ([]byte, str
 	return nil, "", errors.Errorf(`Manifest lists are not supported by "docker-daemon:"`)
 }
 
+type readCloseWrapper struct {
+	io.Reader
+	closeFunc func() error
+}
```

Comment:
There already is a `tarReadCloser` in this file. Modifying that to use this `readCloseWrapper` would be nice.